File: py/rt/final/HitableList.py
class HitableList:
	def __init__(self):
		self.objects = []
		
	def add(self, object):
		self.objects.append(object)

	def hit(self, r, tMin, tMax):		
		hit = None
		for object in self.objects:
			tmpHit = object.hit(r, tMin, tMax)
			if tmpHit != None:
				tMax = tmpHit.t
				hit = tmpHit
		return hit

File: cpp/OpenGL/22_Raytrace/HitableList.cpp
#include "HitableList.h"

#include <iostream>


HitableList::HitableList() 
{	
}

void HitableList::add(std::shared_ptr<const Hitable> hitable) {
	objects.push_back(hitable);
}

const std::optional<HitRecord>HitableList::hit(const Ray& r, const float tMin, const float tMax) const {
	float tClosest = tMax;
	std::optional<HitRecord> hit{};
	for (auto object : objects) {
		std::optional<HitRecord> tmpHit{object->hit(r, tMin, tClosest)};
		if (tmpHit) {
			tClosest = tmpHit->t;
			hit = tmpHit.value();
		}
	}
	return hit;
}

